### ISSUE TITLE
Node: babel not transpiled on when `instrument:false`

### ISSUE THREAD

**Issue by yanivefraim** [wallabyjs/public#717]
I'm having issues with node and babel: when I have `instrument:false` on specific files, those files are not transpiled by babel. I tried removing `instrument:false` from those files and got another issue, because I'm spawning those in a new process, getting an:

 `ReferenceError: $_$wp is not defined` error.
I added an [example repo](https://github.com/yanivefraim/wallaby-node-spawn-process-demo) for this
### Wallaby.js configuration file

```javascript
module.exports = function (wallaby) {
  process.env.NODE_PATH += `:${require('path').join(wallaby.localProjectDir, 'node_modules')}`;
  console.log(wallaby.projectCacheDir);
  return {
    files: [
      {pattern: 'lib/launch.js', instrument: false},
      {pattern: 'lib/index.js', instrument: false},
      {pattern: 'src/**'}
    ],

    tests: [
      {pattern: 'test/**'}
    ],

    compilers: {
      '**/*.js*': wallaby.compilers.babel()
    },

    setup: function () {
      require("babel-polyfill");
    },

    env: {
      type: 'node'
    }
  };
};

```

### Code editor or IDE name and version

Atom v1.x

### OS name and version

OSX

**Comment by ArtemGovorov** [comment]
Thanks for the repo, I have [sent you the PR](https://github.com/yanivefraim/wallaby-node-spawn-process-demo/pull/1). The idea is to just use babel [preprocessor](https://wallabyjs.com/docs/config/preprocessors.html) for those files that can not be instrumented.

**Comment by yanivefraim** [comment]
Thanks for the super-quick reply. It did solve the problem in example repo. Unfortunately, it did not solve the issue in my real world project. I'm still getting `SyntaxError: Unexpected reserved word` for the file that I use `instrument:false`. My fix is:

```js
preprocessors: {
      'src/config.js': file => require('babel-core')
        .transform(file.content, {sourceMap: true, filename: file.path}),
      'index.js': file => require('babel-core')
        .transform(file.content, {sourceMap: true, filename: file.path})
    },
```

When opening the file from `instruments` folder, I see that it is transpiled. I will try to figure this out (BTW - am I using preprocessor for multiple file correctly?)

**Comment by ArtemGovorov** [comment]
> am I using preprocessor for multiple file correctly?

You may simplify it a bit:
```javascript
const babelPreprocessor = file => require('babel-core')
        .transform(file.content, {sourceMap: true, filename: file.path});
...
preprocessors: {
      'src/config.js': babelPreprocessor,
      'index.js': babelPreprocessor
    },
```

**Comment by yanivefraim** [comment]
Thanks. 
The `instrument: false` workaround works, but this is a workaround. What if I want to instrument this code? Can you think of a better way to workaround this?

**Comment by ArtemGovorov** [comment]
It depends on your scenario.

If you have some unit tests that execute the code, then it should run in one process and you should get all the coverage.

If you have some integration tests that execute the code in a different process, then you may set the `instrument: false` (as you are probably testing the fact that you can fork those files, because the logic in those files can be tested without a separate process). Note that wallaby doesn't support cross process code coverage. 

If you have both kind of tests, so can't have `instrument: false` because of the scenario 1 and still need the scenario 2, you may do this:
```javascript
module.exports = function (wallaby) {
  return {
    ...
    setup: function () {
      require("babel-polyfill");

      if (!global.$_$wp) {
        global.$_$wp = global.$_$wpe = global.$_$w = global.$_$wf = () => {};
      }
    },

    ...
  };
};
```

This way if the coverage can be collected, it will be collected, if not - it will not break anything.

**Comment by yanivefraim** [comment]
So I do have a mixed scenario, because I'm working on a universal app. So server code is running on a cluster, using same components that client uses. If I have 'instrument: false' on those components, client (wallaby) functionality will not be complete (no instruments for client components). 

Just making sure: when will `setup` run - for each test? it is like `beforeEach`?

**Comment by ArtemGovorov** [comment]
You may set `instrument: true` and add the noop function assigned to wallaby global instrumentation functions as I have mentioned above (in the `setup` function). This way your client tests that access the code in the same process with have the coverage, and your server tests that fork processes will still work (because for them the instrumentation functions will be stubbed with a noop function).

The `setup` function [gets executed before all of your tests](https://wallabyjs.com/docs/config/bootstrap.html) that run in a particular node process (there may be multiple processes running tests in parallel), so it's more like `beforeAll`.

**Comment by ArtemGovorov** [comment]
@yanivefraim Sorry, you're right, this approach will not work in cluster. 

I have another solution in mind, but I may need some example code to test it. The idea is to monkey-patch the `fork` so that it loads the version of the file from your local project directly when fork, something like:
```javascript
setup: function (w) {
  ...
  const path = require('path');
  const fork = require('child_process').fork;
  require('child_process').fork = function() {
    const localFilePath = path.join(w.localProjectDir, arguments[0]);
    arguments[0] = localFilePath;
    console.log('Replaced ' + arguments[0] + ' with ' + localFilePath);
    return fork.apply(this, arguments);
  };
  ...
}
```

**Comment by yanivefraim** [comment]
Thanks, I will give it a try. 
BTW - I have [this](https://github.com/yanivefraim/wallaby-node-spawn-process-demo) example that we can test.

**Comment by ArtemGovorov** [comment]
Ok, will try the example. You have compiled files in your project's `dist` folder, correct? Because obviously the files that are forked has to be compiled as well.

**Comment by ArtemGovorov** [comment]
Something like this should work:
```javascript
setup: function (w) {
      ...
      const path = require('path');
      const fork = require('child_process').fork;
      require('child_process').fork = function () {
        const localFilePath = arguments[0].replace(
          w.projectCacheDir, 
          path.join(w.localProjectDir, 'YOUR_DIST_FOLDER'));
        arguments[0] = localFilePath;
        return fork.apply(this, arguments);
      };
      ...
    },
```
The example doesn't have the `dist` folder like your real project probably does, but hopefully you get the idea.

**Comment by yanivefraim** [comment]
Not sure that I understand regarding `dist`. I compile files to `dist` using babel using my build process. But, when using wallaby I do not use `dist` folder at all.

**Comment by ArtemGovorov** [comment]
Yes, but in this case you actually want to both use wallaby (for client tests) on certain files and not use it (when those files are forked). What I was suggesting is to run those forked files directly from your `dist` folder. Would that work for you?

**Comment by ArtemGovorov** [comment]
Ok, [here is another solution](https://github.com/yanivefraim/wallaby-node-spawn-process-demo/pull/2) - using a preprocessor, to mock wallaby instrumentation instructions for cases when they can't run (cross process):

```javascript
module.exports = function (wallaby) {
  process.env.NODE_PATH += `:${require('path').join(wallaby.localProjectDir, 'node_modules')}`;

  return {
    files: [
      {pattern: 'lib/*.js', instrument: true}, // <---
      {pattern: 'src/**/*.js'}
    ],

    tests: [
      {pattern: 'test/**/*.js'}
    ],

    compilers: {
      '**/*.js*': wallaby.compilers.babel()
    },

    preprocessors: {
      'lib/*.js': file => '!global.$_$wp && (global.$_$wp = global.$_$wpe = global.$_$w = global.$_$wf = () => {});' + file.content
    },

    env: {
      type: 'node'
    }
  };
};
```

This way files will always be compiled and instrumented by wallaby, and when possible (same process) the coverage will work.

**Comment by yanivefraim** [comment]
Yes! The last solution did the trick. I now have instrument + no need for babel preprocessing. 

10x!!!

**Comment by ArtemGovorov** [comment]
Awesome, thanks for the update!

**Comment by yanivefraim** [comment]
I will update the example, I will be happy to contribute it to wallaby examples... 

I also plan to add another example for universal react app using wallaby (already have it working on our projects, I tend to build a simplified example)

10x again!
